server/model/puzzle.ts

```diff
@@ -127,3 +137,55 @@ export async function addPuzzle(puzzle: PuzzleJson, isPublic = false, pid?: stri
   );
   return pid;
 }
+
+async function isGidAlreadySolved(gid: string) {
+  // Note: This gate makes use of the assumption "one pid per gid";
+  // The unique index on (pid, gid) is more strict than this
+  const {
+    rows: [{count}],
+  } = await pool.query(
+    `
+    SELECT COUNT(*)
+    FROM puzzle_solves
+    WHERE gid=$1
+  `,
+    [gid]
+  );
+  return count > 0;
+}
+
+export async function recordSolve(pid: string, gid: string, timeToSolve: number) {
+  const solved_time = Date.now();
+  const client = await pool.connect();
+
+  // Clients may log a solve multiple times; skip logging after the first one goes through
```

Comment:
I think this was the cause of the stalls, because we early returned without closing client.

Do we actually need this check, or can we just rely on the db to fail if we ever have a (pid,gid) pair already? (It should be super fast as it will have an index on (pid,gid)

Reply:
yeah, you're right. my mistake 😄 .

Minor point of concern:
I'm a bit wary of leaning on the transaction rollback to ensure consistency, I'd rather think of the index as a last line of defense. My wariness comes from
- clients will possibly send a lot of recordSolve (anywhere from 1 to 10 would be typical, i think)
- semantics of exception handling typically imply that it's not a primary code path

Another proposal which I don't know how to implement: (warning -- really bad sql pseudocode)
```
1) insert (gid,pid) into puzzle_solves if not already there
2) set puzzles[pid].solves = count(pid in puzzle_solves)
```

This way, there's no error handling necessary.